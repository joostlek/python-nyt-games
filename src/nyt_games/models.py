"""Models for NYT Games."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date
from typing import Any, Generic, TypeVar

from mashumaro import field_options
from mashumaro.mixins.orjson import DataClassORJSONMixin
from mashumaro.types import SerializationStrategy

T = TypeVar("T")


class OptionalStringSerializationStrategy(SerializationStrategy):
    """Serialization strategy for optional strings."""

    def serialize(self, value: date | None) -> date | None:
        """Serialize optional string."""
        return value

    def deserialize(self, value: str | None) -> date | None:
        """Deserialize optional string."""
        if not value:
            return None
        return date.fromisoformat(value)


@dataclass
class Root(DataClassORJSONMixin, Generic[T]):
    """LatestData model."""

    player: Player[T]


@dataclass
class Player(DataClassORJSONMixin, Generic[T]):
    """LatestData model."""

    stats: T
    user_id: int


@dataclass
class LatestDataStats(DataClassORJSONMixin):
    """LatestData model."""

    wordle: Wordle
    spelling_bee: SpellingBee | None = None


@dataclass
class Connections(DataClassORJSONMixin):
    """Connections model."""

    puzzles_completed: int
    puzzles_won: int
    last_completed: date = field(metadata=field_options(alias="last_played_print_date"))
    current_streak: int
    max_streak: int
    mistakes: dict[str, int]

    @classmethod
    def __pre_deserialize__(cls, d: dict[str, dict[str, Any]]) -> dict[str, Any]:
        """Pre deserialization hook."""
        return d["connections"]


@dataclass
class Wordle(DataClassORJSONMixin):
    """Wordle model."""

    games_played: int = field(metadata=field_options(alias="gamesPlayed"))
    games_won: int = field(metadata=field_options(alias="gamesWon"))
    guesses: dict[str, int]
    current_streak: int = field(metadata=field_options(alias="currentStreak"))
    max_streak: int = field(metadata=field_options(alias="maxStreak"))

    @classmethod
    def __pre_deserialize__(cls, d: dict[str, dict[str, Any]]) -> dict[str, Any]:
        """Pre deserialization hook."""
        return d["legacyStats"]


@dataclass
class SpellingBee(DataClassORJSONMixin):
    """SpellingBee mode."""

    puzzles_started: int
    total_words: int
    total_pangrams: int
    ranks: SpellingBeeRanks


@dataclass
class SpellingBeeRanks(DataClassORJSONMixin):
    """SpellingBeeRanks mode."""

    beginner: int = field(metadata=field_options(alias="Beginner"), default=0)
    good: int = field(metadata=field_options(alias="Good"), default=0)
    good_start: int = field(metadata=field_options(alias="Good Start"), default=0)
    moving_up: int = field(metadata=field_options(alias="Moving Up"), default=0)
    nice: int = field(metadata=field_options(alias="Nice"), default=0)
    solid: int = field(metadata=field_options(alias="Solid"), default=0)


class WordleStats(Root[LatestDataStats]):
    """WordleStats model."""


class ConnectionsStats(Root[Connections]):
    """ConnectionsStats model."""
