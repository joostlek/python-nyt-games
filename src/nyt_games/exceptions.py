"""Asynchronous Python client for NYT Games."""


class NYTGamesError(Exception):
    """Generic exception."""


class NYTGamesAuthenticationError(NYTGamesError):
    """NYT Games authentication exception."""


class NYTGamesConnectionError(NYTGamesError):
    """NYT Games connection exception."""


class NYTGamesParseError(NYTGamesError):
    """NYT Games parse exception."""
