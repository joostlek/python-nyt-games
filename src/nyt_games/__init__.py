"""Asynchronous Python client for NYT Games."""

from nyt_games.exceptions import (
    NYTGamesAuthenticationError,
    NYTGamesConnectionError,
    NYTGamesError,
    NYTGamesParseError,
)
from nyt_games.models import (
    Connections,
    ConnectionsStats,
    LatestDataStats,
    SpellingBee,
    SpellingBeeRanks,
    Wordle,
    WordleStats,
)
from nyt_games.nyt_games import NYTGamesClient

__all__ = [
    "NYTGamesClient",
    "NYTGamesError",
    "NYTGamesConnectionError",
    "NYTGamesAuthenticationError",
    "NYTGamesParseError",
    "LatestDataStats",
    "Connections",
    "SpellingBee",
    "SpellingBeeRanks",
    "WordleStats",
    "ConnectionsStats",
    "Wordle",
]
