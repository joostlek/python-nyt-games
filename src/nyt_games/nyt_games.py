"""Asynchronous Python client for NYT Games."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from importlib import metadata
import socket
from typing import TYPE_CHECKING

from aiohttp import ClientError, ClientResponseError, ClientSession
from yarl import URL

from .exceptions import NYTGamesAuthenticationError, NYTGamesConnectionError
from .models import Connections, ConnectionsStats, LatestDataStats, WordleStats

if TYPE_CHECKING:
    from typing_extensions import Self


VERSION = metadata.version(__package__)


@dataclass
class NYTGamesClient:
    """Main class for handling connections with NYT Games."""

    token: str
    session: ClientSession | None = None
    request_timeout: int = 10
    _close_session: bool = False

    async def _request(self, uri: str, params: dict[str, str] | None = None) -> str:
        """Handle a request to NYT Games."""
        url = URL.build(
            scheme="https",
            host="nytimes.com",
            port=443,
        ).joinpath(uri)

        headers = {
            "User-Agent": f"PythonNYTGames/{VERSION}",
            "Accept": "application/json",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True
        self.session.cookie_jar.update_cookies({"NYT-S": self.token})

        try:
            async with asyncio.timeout(self.request_timeout):
                response = await self.session.get(url, headers=headers, params=params)
        except asyncio.TimeoutError as exception:
            msg = "Timeout occurred while connecting to NYT Games"
            raise NYTGamesConnectionError(msg) from exception
        except (
            ClientError,
            ClientResponseError,
            socket.gaierror,
        ) as exception:
            msg = "Error occurred while communicating with NYT Games"
            raise NYTGamesConnectionError(msg) from exception

        if response.status == 403:
            msg = "Unauthenticated"
            raise NYTGamesAuthenticationError(msg)

        if response.status != 200:
            content_type = response.headers.get("Content-Type", "")
            text = await response.text()
            msg = "Unexpected response from NYT Games"
            raise NYTGamesConnectionError(
                msg,
                {"Content-Type": content_type, "response": text},
            )

        return await response.text()

    async def _get_wordle_stats(self) -> WordleStats:
        response = await self._request("svc/games/state/wordleV2/latests")
        return WordleStats.from_json(response)

    async def get_user_id(self) -> int:
        """Get user identifier."""
        return (await self._get_wordle_stats()).player.user_id

    async def get_latest_stats(self) -> LatestDataStats:
        """Get latest stats."""
        return (await self._get_wordle_stats()).player.stats

    async def get_connections(self) -> Connections | None:
        """Get connections stats."""
        response = await self._request(
            "svc/games/state/connections/latests", {"puzzle_ids": "0"}
        )
        if "player" not in response:
            return None
        return ConnectionsStats.from_json(response).player.stats

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The NYTGamesClient object.

        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.

        """
        await self.close()
