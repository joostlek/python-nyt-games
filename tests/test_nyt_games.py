"""Tests for the client."""

from __future__ import annotations

import asyncio
from typing import TYPE_CHECKING, Any

import aiohttp
from aiohttp import ClientError
from aiohttp.hdrs import METH_GET
from aioresponses import CallbackResult, aioresponses
import pytest

from nyt_games import NYTGamesClient, NYTGamesConnectionError, NYTGamesError
from nyt_games.exceptions import NYTGamesAuthenticationError
from tests import load_fixture
from tests.const import HEADERS, MOCK_URL

if TYPE_CHECKING:
    from syrupy import SnapshotAssertion


async def test_putting_in_own_session(
    responses: aioresponses,
) -> None:
    """Test putting in own session."""
    responses.get(
        f"{MOCK_URL}/svc/games/state/wordleV2/latests",
        status=200,
        body=load_fixture("latest.json"),
    )
    async with aiohttp.ClientSession() as session:
        client = NYTGamesClient(session=session, token="abc")
        await client.get_latest_stats()
        assert client.session is not None
        assert not client.session.closed
        await client.close()
        assert not client.session.closed


async def test_creating_own_session(
    responses: aioresponses,
) -> None:
    """Test creating own session."""
    responses.get(
        f"{MOCK_URL}/svc/games/state/wordleV2/latests",
        status=200,
        body=load_fixture("latest.json"),
    )
    client = NYTGamesClient(token="abc")
    await client.get_latest_stats()
    assert client.session is not None
    assert not client.session.closed
    await client.close()
    assert client.session.closed


async def test_unexpected_server_response(
    responses: aioresponses,
    client: NYTGamesClient,
) -> None:
    """Test handling unexpected response."""
    responses.get(
        f"{MOCK_URL}/svc/games/state/wordleV2/latests",
        status=404,
        headers={"Content-Type": "plain/text"},
        body="Yes",
    )
    with pytest.raises(NYTGamesError):
        await client.get_latest_stats()


async def test_unauthorized(
    responses: aioresponses,
    client: NYTGamesClient,
) -> None:
    """Test handling unauthorized response."""
    responses.get(
        f"{MOCK_URL}/svc/games/state/wordleV2/latests",
        status=403,
        body='{"error": "forbidden"}',
    )
    with pytest.raises(NYTGamesAuthenticationError):
        await client.get_latest_stats()


async def test_timeout(
    responses: aioresponses,
) -> None:
    """Test request timeout."""

    # Faking a timeout by sleeping
    async def response_handler(_: str, **_kwargs: Any) -> CallbackResult:
        """Response handler for this test."""
        await asyncio.sleep(2)
        return CallbackResult(body="Goodmorning!")

    responses.get(
        f"{MOCK_URL}/svc/games/state/wordleV2/latests",
        callback=response_handler,
    )
    async with NYTGamesClient(request_timeout=1, token="abc") as client:
        with pytest.raises(NYTGamesConnectionError):
            await client.get_latest_stats()


async def test_client_error(
    client: NYTGamesClient,
    responses: aioresponses,
) -> None:
    """Test client error."""

    async def response_handler(_: str, **_kwargs: Any) -> CallbackResult:
        """Response handler for this test."""
        raise ClientError

    responses.get(
        f"{MOCK_URL}/measures/current",
        callback=response_handler,
    )
    with pytest.raises(NYTGamesConnectionError):
        await client.get_latest_stats()


@pytest.mark.parametrize(
    "fixture", ["latest.json", "new_account.json", "no_spelling_bee.json"]
)
async def test_get_latest(
    responses: aioresponses,
    client: NYTGamesClient,
    snapshot: SnapshotAssertion,
    fixture: str,
) -> None:
    """Test status call."""
    responses.get(
        f"{MOCK_URL}/svc/games/state/wordleV2/latests",
        status=200,
        body=load_fixture(fixture),
    )
    assert await client.get_latest_stats() == snapshot
    responses.assert_called_once_with(
        f"{MOCK_URL}/svc/games/state/wordleV2/latests",
        METH_GET,
        headers=HEADERS,
        params=None,
    )


async def test_get_connections(
    responses: aioresponses,
    client: NYTGamesClient,
    snapshot: SnapshotAssertion,
) -> None:
    """Test retrieving connections."""
    responses.get(
        f"{MOCK_URL}/svc/games/state/connections/latests?puzzle_ids=0",
        status=200,
        body=load_fixture("connections.json"),
    )
    assert await client.get_connections() == snapshot
    responses.assert_called_once_with(
        f"{MOCK_URL}/svc/games/state/connections/latests",
        METH_GET,
        headers=HEADERS,
        params={"puzzle_ids": "0"},
    )


@pytest.mark.parametrize(
    "fixture", ["new_account_connections.json", "newer_account_connections.json"]
)
async def test_get_connections_new_player(
    responses: aioresponses, client: NYTGamesClient, fixture: str
) -> None:
    """Test retrieving connections."""
    responses.get(
        f"{MOCK_URL}/svc/games/state/connections/latests?puzzle_ids=0",
        status=200,
        body=load_fixture(fixture),
    )
    assert await client.get_connections() is None
    responses.assert_called_once_with(
        f"{MOCK_URL}/svc/games/state/connections/latests",
        METH_GET,
        headers=HEADERS,
        params={"puzzle_ids": "0"},
    )


async def test_get_user_id(
    responses: aioresponses,
    client: NYTGamesClient,
) -> None:
    """Test retrieving user_id."""
    responses.get(
        f"{MOCK_URL}/svc/games/state/wordleV2/latests",
        status=200,
        body=load_fixture("latest.json"),
    )
    assert await client.get_user_id() == 218886794


@pytest.mark.parametrize(
    "fixture",
    [
        "crossword_stats_existing_player.json",
        "crossword_stats_new_player.json",
    ],
)
async def test_crossword_stats(
    responses: aioresponses,
    client: NYTGamesClient,
    snapshot: SnapshotAssertion,
    fixture: str,
) -> None:
    """Test fetching crossword stats."""
    fixture_text = load_fixture(fixture)
    url = f"{MOCK_URL}/svc/crosswords/v3/10781499/stats-and-streaks.json"
    responses.get(
        f"{url}?date_start=1988-01-01&start_on_monday=true",
        body=fixture_text,
    )

    stats = await client.get_crossword_stats()

    assert stats == snapshot
