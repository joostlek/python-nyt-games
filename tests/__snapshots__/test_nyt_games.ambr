# serializer version: 1
# name: test_get_latest
  dict({
    'stats': dict({
      'wordle': dict({
        'current_streak': 1,
        'games_played': 33,
        'games_won': 26,
        'guesses': dict({
          '1': 0,
          '2': 1,
          '3': 4,
          '4': 7,
          '5': 10,
          '6': 4,
          'fail': 7,
        }),
        'last_completed': datetime.date(2024, 9, 20),
        'last_won': datetime.date(2024, 9, 20),
        'max_streak': 5,
      }),
    }),
    'user_id': 218886794,
  })
# ---
